File: SupplyChainFrontend/src/app/services/pricing-rule.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { ApiResponse, PaginatedResponse } from './supplier.service';

// A29-P2-06 — Pricing Rules CRUD + resolve preview, backing PricingRulesController (P2-04).

export type PriceType = 'SELLING' | 'COST' | 'PROMOTIONAL' | 'CONTRACT';

export interface CreatePricingRuleRequest {
  variantUuid: string;
  partnerUuid?: string | null;
  priceType: PriceType;
  minQty?: number | null;
  maxQty?: number | null;
  unitPrice: number;
  currencyId?: string | null;
  effectiveFrom?: string | null;
  effectiveTo?: string | null;
}

export interface UpdatePricingRuleRequest {
  priceType: PriceType;
  minQty?: number | null;
  maxQty?: number | null;
  unitPrice: number;
  currencyId: string;
  effectiveFrom: string;
  effectiveTo?: string | null;
  isActive: boolean;
}

export interface PricingRuleModel {
  uuid: string;
  variantUuid: string;
  variantSku: string;
  productName: string;
  partnerUuid?: string | null;
  priceType: PriceType;
  minQty?: number | null;
  maxQty?: number | null;
  unitPrice: number;
  currencyId: string;
  effectiveFrom: string;
  effectiveTo?: string | null;
  isActive: boolean;
  createdDate: string;
}

export interface PricingRuleListFilter {
  variantUuid?: string;
  partnerUuid?: string;
  priceType?: PriceType;
  isActive?: boolean;
  page?: number;
  pageSize?: number;
}

export interface SalePriceResolution {
  found: boolean;
  unitPrice?: number | null;
  currencyId?: string | null;
  tier?: string | null;
  pricingRuleUuid?: string | null;
}

@Injectable({ providedIn: 'root' })
export class PricingRuleService {
  private readonly baseUrl = `${environment.apiUrl}/pricing-rules`;

  constructor(private http: HttpClient) {}

  createRule(req: CreatePricingRuleRequest): Observable<ApiResponse<string>> {
    return this.http.post<ApiResponse<string>>(this.baseUrl, req);
  }

  updateRule(uuid: string, req: UpdatePricingRuleRequest): Observable<ApiResponse> {
    return this.http.put<ApiResponse>(`${this.baseUrl}/${uuid}`, req);
  }

  deleteRule(uuid: string): Observable<ApiResponse> {
    return this.http.delete<ApiResponse>(`${this.baseUrl}/${uuid}`);
  }

  getRuleById(uuid: string): Observable<ApiResponse<PricingRuleModel>> {
    return this.http.get<ApiResponse<PricingRuleModel>>(`${this.baseUrl}/${uuid}`);
  }

  getRules(filter: PricingRuleListFilter): Observable<ApiResponse<PaginatedResponse<PricingRuleModel>>> {
    let params = new HttpParams();
    if (filter.variantUuid) params = params.set('variantUuid', filter.variantUuid);
    if (filter.partnerUuid) params = params.set('partnerUuid', filter.partnerUuid);
    if (filter.priceType)   params = params.set('priceType', filter.priceType);
    if (filter.isActive !== undefined) params = params.set('isActive', String(filter.isActive));
    params = params.set('page', String(filter.page ?? 1));
    params = params.set('pageSize', String(filter.pageSize ?? 50));
    return this.http.get<ApiResponse<PaginatedResponse<PricingRuleModel>>>(this.baseUrl, { params });
  }

  resolvePrice(variantUuid: string, partnerUuid: string | null, qty: number, date?: string): Observable<ApiResponse<SalePriceResolution>> {
    let params = new HttpParams().set('variantUuid', variantUuid).set('qty', String(qty));
    if (partnerUuid) params = params.set('partnerUuid', partnerUuid);
    if (date) params = params.set('date', date);
    return this.http.get<ApiResponse<SalePriceResolution>>(`${this.baseUrl}/resolve`, { params });
  }
}

File: src/SMS.Modules.Inventory/Services/PricingRuleService.cs
using Microsoft.EntityFrameworkCore;
using SMS.Modules.Inventory.Data;
using SMS.Modules.Inventory.Domain;
using SMS.Modules.Inventory.Models;
using SMS.Shared.Common;
using SMS.Shared.Exceptions;
using SMS.Shared.Pagination;

namespace SMS.Modules.Inventory.Services;

internal sealed class PricingRuleService : IPricingRuleService
{
    private readonly InventoryDbContext _db;
    private readonly ITenantContext _tenantContext;
    private readonly IOrganizationCurrencyService _orgCurrency;

    private static readonly HashSet<string> ValidPriceTypes =
    [
        PricingRuleType.Selling, PricingRuleType.Cost, PricingRuleType.Promotional, PricingRuleType.Contract
    ];

    public PricingRuleService(InventoryDbContext db, ITenantContext tenantContext, IOrganizationCurrencyService orgCurrency)
    {
        _db            = db;
        _tenantContext = tenantContext;
        _orgCurrency   = orgCurrency;
    }

    public async Task<Guid> CreateAsync(CreatePricingRuleRequest req, int createdBy)
    {
        var variant = await _db.ProductVariants.FirstOrDefaultAsync(v => v.Uuid == req.VariantUuid)
            ?? throw new NotFoundException("ProductVariant", req.VariantUuid);

        ValidatePriceType(req.PriceType);
        // A CONTRACT rate that names no partner could never resolve (IPricingService skips tier 1
        // entirely when the caller passes no partner) — refuse it here rather than accept a row
        // that can never apply.
        if (req.PriceType == PricingRuleType.Contract && req.PartnerUuid is null)
            throw new BadRequestException("A CONTRACT price requires a partner.");
        ValidateQtyAndDateRange(req.MinQty, req.MaxQty, req.EffectiveFrom, req.EffectiveTo);

        var currencyId = req.CurrencyId
            ?? await _orgCurrency.GetBaseCurrencyIdAsync(_tenantContext.OrganizationId)
            ?? throw new BadRequestException(
                "Currency is required — this organization has no base currency configured, so it must be supplied explicitly.");

        var entity = new PricingRule
        {
            VariantId     = variant.Id,
            PartnerId     = req.PartnerUuid,
            PriceType     = req.PriceType,
            MinQty        = req.MinQty,
            MaxQty        = req.MaxQty,
            UnitPrice     = req.UnitPrice,
            CurrencyId    = currencyId,
            EffectiveFrom = req.EffectiveFrom?.Date ?? DateTime.UtcNow.Date,
            EffectiveTo   = req.EffectiveTo,
            IsActive      = true,
            CreatedBy     = createdBy,
            CreatedDate   = DateTime.UtcNow
        };

        _db.PricingRules.Add(entity);
        await _db.SaveChangesAsync();
        return entity.Uuid;
    }

    public async Task<bool> UpdateAsync(Guid uuid, UpdatePricingRuleRequest req)
    {
        var entity = await _db.PricingRules.FirstOrDefaultAsync(x => x.Uuid == uuid);
        if (entity is null) return false;

        ValidatePriceType(req.PriceType);
        if (req.PriceType == PricingRuleType.Contract && entity.PartnerId is null)
            throw new BadRequestException("A CONTRACT price requires a partner.");
        ValidateQtyAndDateRange(req.MinQty, req.MaxQty, req.EffectiveFrom, req.EffectiveTo);

        entity.PriceType     = req.PriceType;
        entity.MinQty        = req.MinQty;
        entity.MaxQty        = req.MaxQty;
        entity.UnitPrice     = req.UnitPrice;
        entity.CurrencyId    = req.CurrencyId;
        entity.EffectiveFrom = req.EffectiveFrom.Date;
        entity.EffectiveTo   = req.EffectiveTo;
        entity.IsActive      = req.IsActive;

        await _db.SaveChangesAsync();
        return true;
    }

    // Soft delete, matching ProductsController.DeleteProduct — a rule can be referenced by past
    // price-resolution audit trails, so the row stays; it just stops matching.
    public async Task<bool> DeleteAsync(Guid uuid)
    {
        var entity = await _db.PricingRules.FirstOrDefaultAsync(x => x.Uuid == uuid);
        if (entity is null) return false;

        entity.IsActive = false;
        await _db.SaveChangesAsync();
        return true;
    }

    public async Task<PricingRuleModel?> GetByIdAsync(Guid uuid) =>
        await ProjectedQuery().FirstOrDefaultAsync(x => x.Uuid == uuid);

    public async Task<PaginatedResponse<PricingRuleModel>> GetListAsync(PricingRuleListFilter filter)
    {
        var query = ProjectedQuery();

        if (filter.VariantUuid is { } variantUuid)
            query = query.Where(x => x.VariantUuid == variantUuid);
        if (filter.PartnerUuid is { } partnerUuid)
            query = query.Where(x => x.PartnerUuid == partnerUuid);
        if (!string.IsNullOrWhiteSpace(filter.PriceType))
            query = query.Where(x => x.PriceType == filter.PriceType);
        if (filter.IsActive is { } isActive)
            query = query.Where(x => x.IsActive == isActive);

        query = query.OrderByDescending(x => x.CreatedDate);

        var total    = await query.CountAsync();
        var page     = Math.Max(1, filter.Page);
        var pageSize = Math.Clamp(filter.PageSize, 1, 100);

        var data = await query.Skip((page - 1) * pageSize).Take(pageSize).ToListAsync();

        return new PaginatedResponse<PricingRuleModel>
        {
            Data         = data,
            TotalRecords = total,
            Page         = page,
            PageSize     = pageSize,
            TotalPages   = (int)Math.Ceiling(total / (double)pageSize)
        };
    }

    private IQueryable<PricingRuleModel> ProjectedQuery() =>
        _db.PricingRules.AsNoTracking().Select(x => new PricingRuleModel
        {
            Uuid          = x.Uuid,
            VariantUuid   = x.Variant.Uuid,
            VariantSku    = x.Variant.Sku,
            ProductName   = x.Variant.Product.Name,
            PartnerUuid   = x.PartnerId,
            PriceType     = x.PriceType,
            MinQty        = x.MinQty,
            MaxQty        = x.MaxQty,
            UnitPrice     = x.UnitPrice,
            CurrencyId    = x.CurrencyId,
            EffectiveFrom = x.EffectiveFrom,
            EffectiveTo   = x.EffectiveTo,
            IsActive      = x.IsActive,
            CreatedDate   = x.CreatedDate
        });

    private static void ValidatePriceType(string priceType)
    {
        if (!ValidPriceTypes.Contains(priceType))
            throw new BadRequestException(
                $"'{priceType}' is not a valid price type. Expected one of: {string.Join(", ", ValidPriceTypes)}.");
    }

    private static void ValidateQtyAndDateRange(decimal? minQty, decimal? maxQty, DateTime? from, DateTime? to)
    {
        if (minQty is { } min && maxQty is { } max && min > max)
            throw new BadRequestException("Minimum quantity cannot be greater than maximum quantity.");
        if (from is { } f && to is { } t && f > t)
            throw new BadRequestException("Effective-from date cannot be after effective-to date.");
    }
}
